tests/test_dbmiddleware.py

```diff
@@ -0,0 +1,369 @@
+import logging
+from typing import List
+
+import pytest
+import sqlalchemy
+from fastapi import FastAPI
+from pydantic import BaseModel
+from starlette.applications import Starlette
+from starlette.database import transaction
+from starlette.middleware.database import DatabaseMiddleware
+from starlette.requests import Request
+from starlette.responses import JSONResponse
+from starlette.testclient import TestClient
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+# docker run --name fastapi-postgres -e POSTGRES_PASSWORD=bar
+# -e POSTGRES_USER=foo -e POSTGRES_DB=fastapi -p 5432:5432 -d postgres
+try:
+    DATABASE_URL = "postgresql://foo:bar@localhost/fastapi"
+except KeyError:  # pragma: no cover
+    pytest.skip("DATABASE_URL is not set", allow_module_level=True)
+
+metadata = sqlalchemy.MetaData()
+
+fastapinotes = sqlalchemy.Table(
+    "fastapinotes",
+    metadata,
+    sqlalchemy.Column("id", sqlalchemy.Integer, primary_key=True),
+    sqlalchemy.Column("text", sqlalchemy.String),
+    sqlalchemy.Column("completed", sqlalchemy.Boolean),
+)
+
+starlettenotes = sqlalchemy.Table(
+    "starlettenotes",
+    metadata,
+    sqlalchemy.Column("id", sqlalchemy.Integer, primary_key=True),
+    sqlalchemy.Column("text", sqlalchemy.String),
+    sqlalchemy.Column("completed", sqlalchemy.Boolean),
+)
+
+
+fastapi = FastAPI()
+fastapi.add_middleware(
+    DatabaseMiddleware, database_url=DATABASE_URL, rollback_on_shutdown=True
+)
+
+
+@pytest.fixture(autouse=True, scope="function")
+def create_test_database():
+    engine = sqlalchemy.create_engine(DATABASE_URL)
+    metadata.create_all(engine)
+    yield
+    engine.execute("DROP TABLE fastapinotes")
+    engine.execute("DROP TABLE starlettenotes")
+
+
+class NoteIn(BaseModel):
+    text: str
+    completed: bool
+
+
+class NoteOut(BaseModel):
+    text: str
+    completed: bool
+
+
+class NoteOutJustText(BaseModel):
+    text: str
+
+
+class NoteDB(BaseModel):
```

Comment:
This class is unused

Reply:
yep good catch of a leftover ;)

Reply:
Your PR helped me alot 😄 I also had a watch on [full-stack-fastapi-couchbase](https://github.com/tiangolo/full-stack-fastapi-couchbase). Maybe the idea was to transform an asyncpg row into an object of type NoteDB:

```python
from asyncpg import Record
from starlette.database import DatabaseSession

...

async def read_note(db: DatabaseSession, note_id: int) -> NoteDB:
    query = fastapinotes.select().where(fastapinotes.c.id == note_id)
    row: Record = await request.database.fetchone(query)
    logger.debug("result")
    logger.debug(result)
    return NoteDB(**dict(row))

@fastapi.get("/notes/{note_id}", response_model=NoteOut)
async def read_note_fastapi_starlette(request: Request, note_id: int) -> NoteDB:
    """
    Get a note by id: FastAPI
    """
    db: DatabaseSession = request.database
    return read_note(db, note_id)
```

PyCharm doesn't help much with any type hints, neither on Starlette's request.database nor on SQLAlchemy's query builder 😞